go/teams/seitan_v2.go

```diff
@@ -0,0 +1,201 @@
+}
+
+func (sikey SeitanSIKeyV2) generateKeyPair() (key libkb.NaclSigningKeyPair, err error) {
+	type PrivateKeySeedPayload struct {
+		Stage   string `codec:"stage" json:"stage"`
+		Version int    `codec:"version" json:"version"`
+	}
+
+	payload, err := libkb.MsgpackEncode(PrivateKeySeedPayload{
+		Stage:   "eddsa",
+		Version: 2,
+	})
+	if err != nil {
+		return key, err
+	}
+
+	mac := hmac.New(sha512.New, sikey[:])
+	_, err = mac.Write(payload)
+	if err != nil {
+		return key, err
+	}
+
+	seed := mac.Sum(nil)
+	seed = seed[0:32]
+	pub, priv, err := ed25519.GenerateKey(bytes.NewBuffer(seed))
+	if err != nil {
+		return key, err
+	}
+
+	copy(key.Public[:], pub[:])
+	key.Private = &libkb.NaclSigningKeyPrivate{}
+	copy(key.Private[:], priv[:])
+	return key, nil
+}
+
+func (sikey SeitanSIKeyV2) generatePackedEncryptedKeyWithSecretKey(secretKey keybase1.Bytes32, gen keybase1.PerTeamKeyGeneration, nonce keybase1.BoxNonce, label keybase1.SeitanKeyLabel) (pkey SeitanPKey, encoded string, err error) {
+
+	keyPair, err := sikey.generateKeyPair()
+	if err != nil {
+		return pkey, encoded, err
+	}
+
+	var keyAndLabel keybase1.SeitanKeyAndLabelVersion2
+	keyAndLabel.K = keybase1.SeitanPubKey(keyPair.GetKID().String())
+	keyAndLabel.L = label
+
+	packedKeyAndLabel, err := libkb.MsgpackEncode(keybase1.NewSeitanKeyAndLabelWithV2(keyAndLabel))
+	if err != nil {
+		return pkey, encoded, err
+	}
+	version := uint(2)
+	return packAndEncryptKeyWithSecretKey(secretKey, gen, nonce, packedKeyAndLabel, version)
+}
+
+func (sikey SeitanSIKeyV2) GeneratePackedEncryptedKey(ctx context.Context, team *Team, label keybase1.SeitanKeyLabel) (pkey SeitanPKey, encoded string, err error) {
+	appKey, err := team.SeitanInviteTokenKeyLatest(ctx)
+	if err != nil {
+		return pkey, encoded, err
+	}
+
+	var nonce keybase1.BoxNonce
+	if _, err = rand.Read(nonce[:]); err != nil {
+		return pkey, encoded, err
+	}
+
+	return sikey.generatePackedEncryptedKeyWithSecretKey(appKey.Key, appKey.KeyGeneration, nonce, label)
+}
+
+// "Signature"
+type SeitanSig libkb.NaclSignature
+type SeitanPubKey libkb.NaclSigningKeyPublic
+
+func GenerateSeitanSignatureMessage(uid keybase1.UID, eldestSeqno keybase1.Seqno, inviteID SCTeamInviteID, time keybase1.Time) (payload []byte, err error) {
+	type SigPayload struct {
+		Stage       string         `codec:"stage" json:"stage"`
+		UID         keybase1.UID   `codec:"uid" json:"uid"`
+		EldestSeqno keybase1.Seqno `codec:"eldest_seqno" json:"eldest_seqno"`
+		CTime       keybase1.Time  `codec:"ctime" json:"ctime"`
+		InviteID    SCTeamInviteID `codec:"invite_id" json:"invite_id"`
+		Version     int            `codec:"version" json:"version"`
+	}
+
+	payload, err = libkb.MsgpackEncode(SigPayload{
+		Stage:       "accept",
+		Version:     2,
+		InviteID:    inviteID,
+		UID:         uid,
+		EldestSeqno: eldestSeqno,
+		CTime:       time,
+	})
+	return payload, err
+}
+
+func VerifySeitanSignatureMessage(pubKey SeitanPubKey, msg []byte, sig SeitanSig) bool {
```

Comment:
I think this function should return an `error` on failure and a `nil` on success. That feels a little safer to me.

Reply:
agreed done